Adds a new 'mounts' command to audit sensitive host paths mounts (#322)
Shopify/kubeaudit: docs/auditors/mounts.md

| /var/run/docker.sock      |  Unix socket used to communicate with Docker daemon                                   |
| /                         |  Filesystem's root |
| /etc                      |  Directory that usually contains all system related configurations files         |
| /root                     |  Home directory of the `root` user                                             |
| /var/run/crio/crio.sock   |  Unix socket used to communicate with the CRI-O Container Engine                                                 |
| /home/admin               |  Home directory of the `admin` user        |
| /var/lib/kubelet          |  Directory for Kublet-related configuration                                                             |
| /var/lib/kubelet/pki      |  Directory containing the certificate and private key of the Kublet                                                               |
| /etc/kubernetes           |  Directory containing Kubernetes related configuration              |
| /etc/kubernetes/manifests |  Directory containing manifest of Kubernetes components                                                         |

## Examples

```

> $ kubeaudit mounts -f /Users/j.courtial/dev/go/src/github.com/kubeaudit/auditors/mounts/fixtures/proc-mounted.yml

Reviewer: ```suggestion
$ kubeaudit mounts -f auditors/mounts/fixtures/proc-mounted.yml
```

---------------- Results for ---------------

  apiVersion: v1
  kind: Pod
  metadata:
    name: pod
    namespace: proc-mounted

--------------------------------------------

-- [error] SensitivePathsMounted
   Message: Sensitive path mounted as volume: proc-volume (/proc -> /host/proc, readOnly: false). It should be removed from the container's mounts list.
   Metadata:
      Container: container
      Mount: proc-volume

```

## Explanation

Mounting some sensitive host paths (like `/etc`, `/proc`, or `/var/run/docker.sock`) may allow a container to access sensitive information from the host like credentials or to spy on other workloads' activity.

These sensitive paths should not be mounted.

Example of a resource which **fails** the `mounts` audit:
```yaml
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
        - name: container
          image: scratch
          volumeMounts:
            - mountPath: /host/proc
              name: proc-volume
      volumes:
        - name: proc-volume
          hostPath:
            path: /proc
```

## Override Errors

First, see the [Introduction to Override Errors](/README.md#override-errors).

> The override identifier has the format `allow-host-path-mount--[mount name]` which allows for each mount to be individually overridden.

Reviewer: I think the double dash should be a single dash to match the code and examples. Honestly, I kind of like the double dash better but single dash would be consistent with capabilities overrides... What do you think?

Also, I saw some logic to convert the volume name to override label (lowercase and replace underscores with dashes). Maybe you did that to be consistent with the capabilities auditor, but if the character set for volume names is allowed in labels I wouldn't be opposed to just using the volume name in the override label without any conversion. For example if the volume name is `my_Volume` then the override label would be `allow-host-path-mount--my_Volume`. What do you think? (I don't actually know if `my_Volume` is a valid volume name but just for argument sake 😄 )

Author: Totally right for both. The double dash was a typo. Regarding the volume name, indeed let's keep things simple and use the volume name as it is 👍

---------------- Results for ---------------

  apiVersion: apps/v1beta2
  kind: Deployment
  metadata:
    name: deployment
    namespace: example-namespace

--------------------------------------------

-- [error] SensitivePathsMounted
   Message: Sensitive path mounted as volume: etc-volume (/etc -> /host/etc, readOnly: false). It should be removed from the container's mounts list.
   Metadata:
      Container: container
      Mount: etc-volume

-- [error] SensitivePathsMounted
   Message: Sensitive path mounted as volume: docker-socket-volume (/var/run/docker.sock -> /var/run/docker.sock, readOnly: false). It should be removed from the container's mounts list.
   Metadata:
      Container: container
      Mount: docker-socket-volume
```

### Example with Custom Paths List

A custom paths list can be provided as a comma separated value list of paths using the `--paths` flag. These are the host paths you'd like to have kubeaudit raise an error when they are mounted in a container.

`manifest.yaml` (example manifest)

```yaml
volumes:
  - name: etc-volume
    hostPath:
      path: /etc
  - name: docker-socket-volume
    hostPath:
      path: /var/run/docker.sock
```

```shell

> $ kubeaudit capabilities --path "/etc,/var/run/docker.sock" -f "manifest.yaml"

Reviewer: ```suggestion
$ kubeaudit capabilities --paths "/etc,/var/run/docker.sock" -f "manifest.yaml"
```

Author: Good catch. Also the `capabilities` instead of `mounts` ... Copy/paste 🙄

`config.yaml`

```yaml
---
auditors:
  mounts:
    paths: ["/etc", "/var/run/docker.sock"]
```

`manifest.yaml`

```yaml
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: deployment
  namespace: example-namespace
spec:
  template:
    spec:
      containers:
        - name: container
          image: scratch
          volumeMounts:
            - mountPath: /host/etc
              name: etc-volume
            - mountPath: /var/run/docker.sock
              name: docker-socket-volume
      volumes:
        - name: etc-volume
          hostPath:
            path: /etc
        - name: docker-socket-volume
          hostPath:
            path: /var/run/docker.sock
```

```shell
$ kubeaudit mounts --kconfig "config.yaml" -f "manifest.yaml"

---------------- Results for ---------------

  apiVersion: apps/v1beta2
  kind: Deployment
  metadata:
    name: deployment
    namespace: example-namespace

--------------------------------------------

-- [error] SensitivePathsMounted
   Message: Sensitive path mounted as volume: etc-volume (/etc -> /host/etc, readOnly: false). It should be removed from the container's mounts list.
   Metadata:
      Container: container

>       Mount: etc-volume

Reviewer: Do you mind adding the other metadata fields here?

#### Default sensitive host paths list

| Host path                 | Description                                                                              |
| :------------------------ | :--------------------------------------------------------------------------------------- |
| /proc                     |  Pseudo-filesystem which provides an interface to kernel data structures                                                  |
| /var/run/docker.sock      |  Unix socket used to communicate with Docker daemon                                   |
| /                         |  Filesystem's root |
| /etc                      |  Directory that usually contains all system related configurations files         |
| /root                     |  Home directory of the `root` user                                             |
| /var/run/crio/crio.sock   |  Unix socket used to communicate with the CRI-O Container Engine                                                 |
| /home/admin               |  Home directory of the `admin` user        |
| /var/lib/kubelet          |  Directory for Kublet-related configuration                                                             |

> | /var/lib/kubelet/pki      |  Directory containing the certificate and private key of the Kublet                                                               |

Reviewer 2: tiny thing:
```suggestion
| /var/lib/kubelet/pki      |  Directory containing the certificate and private key of the kubelet                                                               |
```

| /etc/kubernetes           |  Directory containing Kubernetes related configuration              |
| /etc/kubernetes/manifests |  Directory containing manifest of Kubernetes components                                                         |

## Examples

```
$ kubeaudit mounts -f auditors/mounts/fixtures/proc-mounted.yml

---------------- Results for ---------------

  apiVersion: v1
  kind: Pod
  metadata:
    name: pod
    namespace: proc-mounted

--------------------------------------------

-- [error] SensitivePathsMounted
   Message: Sensitive path mounted as volume: proc-volume (/proc -> /host/proc, readOnly: false). It should be removed from the container's mounts list.
   Metadata:
      Container: container
      MountName: proc-volume
      MountPath: /host/proc
      MountReadOnly: false
      MountVolume: proc-volume
      MountVolumeHostPath: /proc

```

### Example with Config File

> A custom paths list can be provided in the config file. See [docs](docs/all.md) for more information. These are the host
> paths you'd like to have kubeaudit raise an error when they are mounted in a container.
> 

Reviewer 2: ```suggestion
If you don't want kubeaudit to raise errors for all the paths in the default list (`DefaultSensitivePaths`), you can provide a custom paths list in the config file. See [docs](docs/all.md) for more information. That way kubeaudit will only raise errors for those specific paths listed in the config file.
```

Reviewer 2: something along these lines... Because what happens here is that it overrides the default list, so it will only wtach the paths I specify

`config.yaml`

```yaml
---
auditors:
  mounts:
    paths: ["/etc", "/var/run/docker.sock"]

> ```

Reviewer 2: When I tested locally, I also had to turn this auditor on by setting `mounts`to true, like so:

```yaml
enabledAuditors:
  mounts: true
 ```

Reviewer 2: Another suggestion, to make it more explicit (completely optional). Instead of paths, how about `pathsToWatch`?

Author: I went for the simple `paths` to be more aligned with other auditors that also use simple key (`add`, `image`, `cpu`)  that are also the same as the CLI options. But we can totally switch to `pathsToWatch` if you find it better.

Reviewer 2: I see. For the capabilities auditor, however, we add capabilities to an allowAddList  so `kubeaudit` won't raise errors for those.  So it made me think that paths would have a similar effect - listed paths would be the ones `kubeaudit` would ignore. However it's the opposite - `kubeaudit` will only watch for those specific paths, everything else will be ignored. That's why I thought of a more explicit name... what do you think?

Author: Ok understood. I find the phrasing a bit weird however :) I was about to suggest `sensitivesPaths` but I guess the issue is the same, it may not be quite clear that it work as a deny-list and not as an allow-list. So if you prefer, yes we can go for `pathsToWatch`

Reviewer 2: yeah... let's leave it like then, then. And hopefully the documentation will make it clear! 😁

Reviewer 2: or maybe `denyPathsList`? :P

Reviewer 2: @jcbbc can you please change it to `denyPathsList`? with that, we're good to go! thanks! 🙏

Author: Sorry, I missed your previous comment. I've made the change.

Reviewer 2: Thanks a lot, @jcbbc!! :)

`manifest.yaml`

```yaml
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: deployment
  namespace: example-namespace
spec:
  template:
    spec:
      containers:
        - name: container
          image: scratch
          volumeMounts:
            - mountPath: /host/etc
              name: etc-volume
            - mountPath: /var/run/docker.sock
              name: docker-socket-volume
      volumes:
        - name: etc-volume
          hostPath:
            path: /etc
        - name: docker-socket-volume
          hostPath:
            path: /var/run/docker.sock
```

```shell

> $ kubeaudit mounts --kconfig "config.yaml" -f "manifest.yaml"

Reviewer 2: Passing in a config file only works with `all` and `autofix`(https://github.com/Shopify/kubeaudit#configuration-file)

```suggestion
$ kubeaudit all --kconfig  "config.yaml" -f "manifest.yaml"
```

### Flags

| Short   | Long              | Description                                                          | Default                                                                  |
| :------ | :---------------- | :------------------------------------------------------------------- | :----------------------------------------------------------------------- |

> | -s      | --denyPathsList   | List of sensitive paths that shouldn't be mounted.                   | [default sensitive host paths list](#Default-sensitive-host-paths-list)  |

Reviewer 2: I believe we can go for `deny-paths-list` so the flag has the same format as [the others](https://github.com/Shopify/kubeaudit/blob/8be8c73b59a5cac5d1e5178d18f721838ba6b119/cmd/commands/capabilities.go#L12)

Author: Oh yes good catch. Also, I was wondering, do we replace `-s` by `-d` for the `deny` part of the argument?

Reviewer 2: I guess it'd be a good idea to have `-d` as a shorthand, @jcbbc!

```

```shell

> $ kubeaudit mounts --denyPathsList "/etc,/var/run/docker.sock" -f "manifest.yaml"

Reviewer 2: if we change the flag, then this will be:

```$ kubeaudit mounts --deny-paths-list "/etc,/var/run/docker.sock" -f "manifest.yaml"```
